controllers/apps/operations/upgrade.go

```diff
@@ -181,3 +159,52 @@ func (u upgradeOpsHandler) getClusterComponentVersionMap(ctx context.Context,
 	}
 	return components, nil
 }
+
+func (u upgradeOpsHandler) getCompOpsListByClusterVersion(opsRes *OpsResource) []appsv1alpha1.ComponentOps {
+	var compOpsList []appsv1alpha1.ComponentOps
+	for _, v := range opsRes.Cluster.Spec.ComponentSpecs {
+		compOpsList = append(compOpsList, appsv1alpha1.ComponentOps{ComponentName: v.Name})
+	}
+	for _, v := range opsRes.Cluster.Spec.ShardingSpecs {
+		compOpsList = append(compOpsList, appsv1alpha1.ComponentOps{ComponentName: v.Name})
+	}
+	return compOpsList
+}
+
+// getUpgradeComponentDefMap gets the desired componentDefinition map that is intended to be upgraded.
+func (u upgradeOpsHandler) getUpgradeComponentDefMap(reqCtx intctrlutil.RequestCtx,
```

Comment:
This method name get... reads like it's a read-only method. However, it mutate each ComponentDefinition by update image.
So the method name is misleading, should be rename to getUpgradedComponentDefAndUpdateImageWithNewServiceVersion.

Although the method name is much looooonger, its side effect is clearer from reading the name.

Reply:
Or split it into two functions:
One is getComponentDefMap, the other is patchComponentDefWithNewServiceVersion